### ISSUE TITLE
[css-overflow-5] Scrolling to unreachable scroll aligned marker positions

### ISSUE THREAD

**Issue by flackr** [w3c/csswg-drafts#11165]
There are many common circumstances in which the scroll targets of scroll markers are in unreachable positions. For example, this occurs with any list with `scroll-align: center` where the targeted items are significantly smaller than the scrollport.

Demo page: https://flackr.github.io/carousel/examples/prototypes/active-marker/#unreachable
![Screenshot 2024-11-07 at 10 22 15 AM](https://github.com/user-attachments/assets/8693155e-5cdd-4117-8528-f6a72057211b)

Items 1 through 3 cannot get to their scroll aligned positions. Similarly the last 3 items at the least can never reach their scroll aligned position. In #10738 I have proposed a mechanism by which explicitly scrolling to an item would result in it being active even if it is not possible to scroll to. This ensures that at least clicking on the markers would result in them being active and moving to them via keyboard would work. However, as soon as you scroll, it would jump to the aligned item.

Without adding special behavior to avoid this, we would need to decide between one of 2 behaviors, as shown in https://flackr.github.io/carousel/examples/prototypes/active-marker/#unreachable. Either,
* We always use the normal selection behavior, starting with item 4 active, or
* We consider being at the beginning of the scroller as having the first target active and scrolling to the end as having the last item active. This results in a significant jump in the active marker once you begin scrolling.

As mentioned in the linked issue #10738, a developer, authors can avoid these unreachable target situations by:
* adding padding to the edges of their scrolling element, or
* ensuring their scroll marker elements are scrollport sized, or
* enabling infinite / cyclical scrolling as in #5411

We could however think on ideas for how we might expect to be able to select the items outside of the reachable range. I've listed a few potential options with demos below:

Options:
1. Add additional virtual scrolling ([demo](https://flackr.github.io/carousel/examples/prototypes/active-marker/#virtual-scroll)). Naively, we could extend the scrollable range to allow virtually scrolling between the items whose aligned positions is not reachable. This seems unintuitive and confusing.
2. We could add additional real scrolling into the extra content as in #7885. This seems like it could be confusing to do automatically as the scroll range would not match the content range, but it is an option that has come up and adding padding matches the way this is typically solved in practice.
3. Distribute scroll proportionally ([demo](https://flackr.github.io/carousel/examples/prototypes/active-marker/#proportional)). This would mean that we'd map the scrollable range to the proportion of the scroll aligned layout ranges including the unreachable positions. As a result, you'd be able to reach all items, however, the active item would often be different than the visually aligned item.
4. Distribute scroll leading up to first reachable item ([demo](https://flackr.github.io/carousel/examples/prototypes/active-marker/#distribute-ends)). In this case, we find the first item that is at a positive scroll offset (and similarly the last reachable item before the end of scroll) and map the scroll before you reach the first item and after you pass the last between the remaining unreachable items. This preserves the active item matching the aligned item once you reach the first aligned item. It does degenerate a bit to the current situation if your first aligned item is very close to the start of scroll as shown in the second example. 
5. Distribute some amount of the initial and ending scroll ([demo](https://flackr.github.io/carousel/examples/prototypes/active-marker/#distribute-page)). In this case, we ensure that the amount of scroll distance we distribute is non-trivial. In the prototype, we distribute the scroll up to the first item that is more than half a scrollport away from being active. This is similar to option 3 except avoids the edge case where you can immediately jump to the aligned item. Similar to option 2 though, this means that until you reach that amount of scroll the active item does not match the visually aligned item.

I think my preference would be if we decide to do this, to adopt something similar to option 5 but with a greatly reduced range over which we traverse the unreachable markers.

**Comment by flackr** [comment]
Note that developers currently have all of these same problems if they build something using scroll snap and imply selection based on the snapped items. We may want to consider whether we should apply similar changes to snap point selection.

**Comment by flackr** [comment]
Option 1 is yes. The others are different in that they're not changing the layout.

**Comment by tabatkins** [comment]
I definitely think this is a problem to solve, but I hate all the presented options. ^_^

1. Virtual scrolling is confusing; a portion of the scrollbar doesn't cause any motion.
2. *Real* bonus scrolling is awful-looking in demos like this, where there's a large distance between the ideal center and the 0 scroll position. (Especially on my widescreen monitor, phew.)
3. Proportional redistribution screws up the entire range. Just *clearly* wrong in practice, as only 8 and 9 actually register themselves when snapped.
4. Just redistributing up to the next point has unpredictable effects, since the distance to the next exact snap point might be very small (maybe even zero). But when the distance is reasonably-sized, this does look okay, and it's correct for all of the range that *is* alignable.
5. The "redistribute up to half a page" seems to *also* misalign the entire range; the alignable items 4-8 register as something below their actual number, while 9-13 all register as something above their actual number. *Nothing* is correctly aligned. But I suppose if there were even more items they'd eventually align?

So I think a more subtle take on 5 would be fine; rather than half a page, perhaps just the next alignable snap point at least some fixed predetermined distance away. We could also apply this non-linearly (a cubic ramp?) so it still spends most of its time selecting the items near the alignable point, rather than ones close to the start, looking *mostly* correct while still visiting all the items as you scroll.

**Comment by flackr** [comment]
Thanks @tabatkins. I agree, with your take on all of the options (and separated the padding option to match your list), and that I think option 5 but with a significantly reduced distance over which it spreads the edge scroll may be the best among them. I've updated the demo so that it only uses 1/8 of the scrollport size so that we can play with it and get a feel for it. I'm also happy to experiment with non-linear distribution.

As for the relation to scroll snap, I think it's probably okay that this won't align with scroll snap in the edge cases.

I think the proposed resolution here for now would be to include option 5 as part of the algorithm in #10917. I suspect if/when we did make this normative it may be allowed but not required.

**Comment by tabatkins** [comment]
Yeah, 1/8 of a page feels pretty good imo.

**Comment by johannesodland** [comment]
There might not be a one-size-fits-all solution for every use case.

For example, in image galleries embedded in articles, the second option might be useful.
These galleries might take up the full width of the article and use styling or scroll-driven animations to highlight the snapped item while reducing focus on the others – for instance by adjusting opacity, filters or transforms.

In such cases it's crucial that the snapped element appears near the text column, rather than being offset to one side of the article. Adding real scrolling to ensure that the snap point is reachable would solve this use case.

**Comment by tabatkins** [comment]
True, tho that's not inconsistent with option 5. We could go with 5 as the behavior, and then *separately* have the ability to auto-add padding to the scroll container to allow all possible snap targets to be alignable. If you turn on that second ability, it negates option 5, as there's nothing unscrollable.
